src/docfx/lib/markdown/ExtractTitle.cs

```diff
@@ -14,27 +15,67 @@ namespace Microsoft.Docs.Build
 {
     internal static class ExtractTitle
     {
+        [ThreadStatic]
+        private static bool s_headingFound = false;
+
+        [ThreadStatic]
+        private static bool s_waitingForInclusionHeading = false;
+
         public static MarkdownPipelineBuilder UseExtractTitle(this MarkdownPipelineBuilder builder)
         {
             return builder.Use(document =>
             {
-                var h1 = GetHeadingBlock(document);
+                var firstBlock = GetAfterYamlFirstNonCommentsBlock(document);
+                var heading = firstBlock as HeadingBlock;
 
-                if (h1 != null && h1.Level == 1)
+                if (InclusionContext.IsInclude)
                 {
-                    Markup.Result.HtmlTitle = RenderTitle(h1);
-                    document.Remove(h1);
+                    if (!s_headingFound && s_waitingForInclusionHeading)
+                    {
+                        if (heading != null && heading.Level == 1)
+                        {
+                            Markup.Result.HtmlTitle = RenderTitle(heading);
+                            document.Remove(heading);
+                            s_headingFound = true;
```

Comment:
Also s_waitingForInclusionHeading = false?

Reply:
does that matter?

---
In reply to: [212188315](https://github.com/dotnet/docfx/pull/3266#discussion_r212188315) [](ancestors = 212188315)

Reply:
It does not matter for the logic, just the status will be more clear. Have approved.

Reply:
good idea

---
In reply to: [212189030](https://github.com/dotnet/docfx/pull/3266#discussion_r212189030) [](ancestors = 212189030)